File: Web/src/lib/plans/triathlon-time.ts
import { RaceType } from '@/generated/prisma/browser';
import { calculateVdot } from '@/lib/metrics/vdot';
import { estimateSwimPaceFromVdot } from './swim-pace';
import { estimateBikeFtpFromVdot } from './bike-zones';

interface TriathlonTimeInput {
    vdot: number;
    raceType: RaceType;
    customSwimDistM?: number;
    customBikeDistM?: number;
    customRunDistM?: number;
}

interface TriathlonTimeResult {
    totalSeconds: number;
    swimSeconds: number;
    bikeSeconds: number;
    runSeconds: number;
    t1Seconds: number;
    t2Seconds: number;
}

interface TriathlonTimeProjection {
    optimal: TriathlonTimeResult;
    projected: TriathlonTimeResult;
    conservative: TriathlonTimeResult;
}

const TRI_RACE_SWIM: Partial<Record<RaceType, number>> = {
    SPRINT_TRI: 750,
    OLYMPIC_TRI: 1500,
    HALF_IRONMAN: 1900,
    FULL_IRONMAN: 3800,
    CUSTOM_TRI: 1500,
};

const TRI_RACE_BIKE: Partial<Record<RaceType, number>> = {
    SPRINT_TRI: 20000,
    OLYMPIC_TRI: 40000,
    HALF_IRONMAN: 90000,
    FULL_IRONMAN: 180000,
    CUSTOM_TRI: 40000,
};

const TRI_RACE_RUN: Partial<Record<RaceType, number>> = {
    SPRINT_TRI: 5000,
    OLYMPIC_TRI: 10000,
    HALF_IRONMAN: 21097,
    FULL_IRONMAN: 42195,
    CUSTOM_TRI: 10000,
};

type TransitionTimes = { t1: number; t2: number };

const TRI_TRANSITIONS: Partial<Record<RaceType, TransitionTimes>> = {
    SPRINT_TRI: { t1: 120, t2: 90 },
    OLYMPIC_TRI: { t1: 150, t2: 120 },
    HALF_IRONMAN: { t1: 330, t2: 210 },
    FULL_IRONMAN: { t1: 510, t2: 360 },
};

const DEFAULT_TRANSITION: TransitionTimes = { t1: 150, t2: 120 };

function predictTimeForDistance(vdot: number, distanceM: number): number {
    let low = 600;
    let high = 18000;
    for (let i = 0; i < 50; i++) {
        const mid = (low + high) / 2;
        const testVdot = calculateVdot({ distance: distanceM, timeSeconds: mid });
        if (Math.abs(testVdot - vdot) < 0.01) return Math.round(mid);
        if (testVdot > vdot) low = mid;
        else high = mid;
    }
    return Math.round((low + high) / 2);
}

function bikePowerToSpeed(watts: number): number {
    return Math.pow(watts / 0.38, 1 / 3);
}

function computeTriathlonSplits(vdot: number, swimDistM: number, bikeDistM: number, runDistM: number, transitions: TransitionTimes): TriathlonTimeResult {
    const swimSeconds = (estimateSwimPaceFromVdot(vdot) * swimDistM) / 100;
    const ftp = estimateBikeFtpFromVdot(vdot);
    const bikeSeconds = bikeDistM / bikePowerToSpeed(ftp * 0.75);
    const runSeconds = predictTimeForDistance(vdot, runDistM);
    const totalSeconds = swimSeconds + bikeSeconds + runSeconds + transitions.t1 + transitions.t2;
    return {
        totalSeconds: Math.round(totalSeconds),
        swimSeconds: Math.round(swimSeconds),
        bikeSeconds: Math.round(bikeSeconds),
        runSeconds: Math.round(runSeconds),
        t1Seconds: transitions.t1,
        t2Seconds: transitions.t2,
    };
}

export function estimateTriathlonTime(input: TriathlonTimeInput): TriathlonTimeProjection | null {
    if (input.vdot <= 0) return null;

    const fallbackRace = input.raceType as RaceType;
    const swimDistM = (input.customSwimDistM && input.customSwimDistM > 0) ? input.customSwimDistM : (TRI_RACE_SWIM[fallbackRace] ?? TRI_RACE_SWIM.SPRINT_TRI!);
    const bikeDistM = (input.customBikeDistM && input.customBikeDistM > 0) ? input.customBikeDistM : (TRI_RACE_BIKE[fallbackRace] ?? TRI_RACE_BIKE.SPRINT_TRI!);
    const runDistM = (input.customRunDistM && input.customRunDistM > 0) ? input.customRunDistM : (TRI_RACE_RUN[fallbackRace] ?? TRI_RACE_RUN.SPRINT_TRI!);
    const transitions = TRI_TRANSITIONS[fallbackRace] ?? DEFAULT_TRANSITION;

    const optimal = computeTriathlonSplits(input.vdot, swimDistM, bikeDistM, runDistM, transitions);

    const isLongCourse = fallbackRace === 'HALF_IRONMAN' || fallbackRace === 'FULL_IRONMAN';
    const projRunMult = isLongCourse ? 1.15 : 1.10;
    const consRunMult = isLongCourse ? 1.25 : 1.20;

    const projSwim = Math.round(optimal.swimSeconds * 1.05);
    const projBike = Math.round(optimal.bikeSeconds * 1.08);
    const projRun = Math.round(optimal.runSeconds * projRunMult);
    const projected: TriathlonTimeResult = {
        totalSeconds: projSwim + projBike + projRun + optimal.t1Seconds + optimal.t2Seconds,
        swimSeconds: projSwim,
        bikeSeconds: projBike,
        runSeconds: projRun,
        t1Seconds: optimal.t1Seconds,
        t2Seconds: optimal.t2Seconds,
    };

    const consSwim = Math.round(optimal.swimSeconds * 1.10);
    const consBike = Math.round(optimal.bikeSeconds * 1.15);
    const consRun = Math.round(optimal.runSeconds * consRunMult);
    const conservative: TriathlonTimeResult = {
        totalSeconds: consSwim + consBike + consRun + optimal.t1Seconds + optimal.t2Seconds,
        swimSeconds: consSwim,
        bikeSeconds: consBike,
        runSeconds: consRun,
        t1Seconds: optimal.t1Seconds,
        t2Seconds: optimal.t2Seconds,
    };

    return { optimal, projected, conservative };
}

export type { TriathlonTimeInput, TriathlonTimeResult, TriathlonTimeProjection };

File: android/app/src/main/java/com/runflow2/app/domain/plan/TriathlonTime.kt
package com.runflow2.app.domain.plan

import com.runflow2.app.core.math.VdotMath
import com.runflow2.app.domain.model.RaceType
import kotlin.math.pow
import kotlin.math.roundToInt

/** One swim → T1 → bike → T2 → run split of a projected triathlon finish. */
data class TriathlonSplits(
    val totalSeconds: Int,
    val swimSeconds: Int,
    val bikeSeconds: Int,
    val runSeconds: Int,
    val t1Seconds: Int,
    val t2Seconds: Int,
)

data class TriathlonLegs(val swimM: Double, val bikeM: Double, val runM: Double)

data class TriathlonTimeProjection(
    val optimal: TriathlonSplits,
    val projected: TriathlonSplits,
    val conservative: TriathlonSplits,
    val legs: TriathlonLegs,
)

/**
 * Port of Web/src/lib/plans/triathlon-time.ts (L29-137) — projects a total
 * finish time (and per-leg splits) from a run VDOT. Used by the plan wizard
 * so triathlon targets show the whole race, not just the run leg.
 */
object TriathlonTimeEstimator {

    private val SWIM_M: Map<String, Double> = mapOf(
        "SPRINT_TRI" to 750.0, "OLYMPIC_TRI" to 1500.0, "HALF_IRONMAN" to 1900.0,
        "FULL_IRONMAN" to 3800.0, "CUSTOM_TRI" to 750.0,
    )
    private val BIKE_M: Map<String, Double> = mapOf(
        "SPRINT_TRI" to 20000.0, "OLYMPIC_TRI" to 40000.0, "HALF_IRONMAN" to 90000.0,
        "FULL_IRONMAN" to 180000.0, "CUSTOM_TRI" to 20000.0,
    )
    private val RUN_M: Map<String, Double> = mapOf(
        "SPRINT_TRI" to 5000.0, "OLYMPIC_TRI" to 10000.0, "HALF_IRONMAN" to 21097.0,
        "FULL_IRONMAN" to 42195.0, "CUSTOM_TRI" to 5000.0,
    )

    private data class Transitions(val t1: Int, val t2: Int)

    private val TRANSITIONS: Map<String, Transitions> = mapOf(
        "SPRINT_TRI" to Transitions(120, 90),
        "OLYMPIC_TRI" to Transitions(150, 120),
        "HALF_IRONMAN" to Transitions(330, 210),
        "FULL_IRONMAN" to Transitions(510, 360),
    )
    private val DEFAULT_TRANSITION = Transitions(150, 120)

    fun estimate(
        vdot: Double,
        raceType: RaceType,
        customSwimDistM: Double? = null,
        customBikeDistM: Double? = null,
        customRunDistM: Double? = null,
    ): TriathlonTimeProjection? {
        if (vdot <= 0.0) return null

        val key = raceType.name
        val swim = customSwimDistM?.takeIf { it > 0 } ?: (SWIM_M[key] ?: SWIM_M.getValue("SPRINT_TRI"))
        val bike = customBikeDistM?.takeIf { it > 0 } ?: (BIKE_M[key] ?: BIKE_M.getValue("SPRINT_TRI"))
        val run = customRunDistM?.takeIf { it > 0 } ?: (RUN_M[key] ?: RUN_M.getValue("SPRINT_TRI"))
        val t = TRANSITIONS[key] ?: DEFAULT_TRANSITION

        val optimal = splits(vdot, swim, bike, run, t)

        // longer courses degrade the run more (triathlon-time.ts L108-110)
        val isLongCourse = (swim + bike + run) / 1000.0 >= 100.0
        val projRunMult = if (isLongCourse) 1.15 else 1.10
        val consRunMult = if (isLongCourse) 1.25 else 1.20

        fun scaled(swimMult: Double, bikeMult: Double, runMult: Double): TriathlonSplits {
            val s = (optimal.swimSeconds * swimMult).roundToInt()
            val b = (optimal.bikeSeconds * bikeMult).roundToInt()
            val r = (optimal.runSeconds * runMult).roundToInt()
            return TriathlonSplits(s + b + r + t.t1 + t.t2, s, b, r, t.t1, t.t2)
        }

        val projected = scaled(1.05, 1.08, projRunMult)
        val conservative = scaled(1.10, 1.15, consRunMult)
        return TriathlonTimeProjection(optimal, projected, conservative, TriathlonLegs(swim, bike, run))
    }

    private fun splits(
        vdot: Double,
        swimDistM: Double,
        bikeDistM: Double,
        runDistM: Double,
        t: Transitions,
    ): TriathlonSplits {
        val swimSec = estimateSwimPaceFromVdot(vdot) * swimDistM / 100.0
        val ftp = estimateBikeFtpFromVdot(vdot)
        val bikeSec = bikeDistM / bikePowerToSpeed(ftp * 0.75)
        val runSec = VdotMath.predictTimeSec(vdot, runDistM) ?: runDistM / 3.0
        // round each leg, then total the rounded legs so the displayed
        // splits always add up exactly (web rounds independently; ≤1s drift)
        val swimR = swimSec.roundToInt()
        val bikeR = bikeSec.roundToInt()
        val runR = runSec.roundToInt()
        return TriathlonSplits(
            totalSeconds = swimR + bikeR + runR + t.t1 + t.t2,
            swimSeconds = swimR,
            bikeSeconds = bikeR,
            runSeconds = runR,
            t1Seconds = t.t1,
            t2Seconds = t.t2,
        )
    }

    /** triathlon-time.ts `bikePowerToSpeed` (L77-79): m/s at ~0.38 CdA-ish drag constant. */
    private fun bikePowerToSpeed(watts: Double): Double = (watts / 0.38).pow(1.0 / 3.0)
}
